### PR TITLE
Domains: add/uk transfer out

### PR DESCRIPTION
**Pull request by hewsut**
.uk domains have a separate ("push-based", via setting a special registrar tag) domain transfer flow. This PR adds support. See the video for a demonstration:

Success flow:
success

Fail flow:
fail

~D9676-code on the back~ (committed)

### CODE REVIEW

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	setSuggestionsRef = ref => ( this.suggestionsRef = ref );
+
+	hideSuggestions = () => this.setState( { query: '' } );
+
+	render() {
+		const { selectedDomainName, translate } = this.props;
+
+		return (
+			<div>
+				<SectionHeader label={ translate( 'Transfer Domain To Another Registrar' ) } />
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer to another registrar' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+			</Dialog>
+		);
+	}
+
+	render() {
+		const { translate } = this.props;
+
+		return (
+			<div>
+				<SectionHeader label={ translate( 'Transfer Domain To Another Registrar' ) } />
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer to another registrar' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+			{ action: 'cancel', label: translate( 'Cancel' ) },
+			{ action: 'submit', label: translate( 'Yes, Submit!' ), isPrimary: true },
+		];
+
+		return (
+			<Dialog
+				isVisible={ this.state.showDialog }
+				buttons={ buttons }
+				onClose={ this.onCloseDialog }>
+				<h1>{ translate( 'Transfer Confirmation' ) }</h1>
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer confirmation required' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+			</div>
+		);
+	}
+
+	renderDialog() {
+		const { selectedDomainName, translate } = this.props;
+		const { selectedRegistrar } = this.state;
+		const buttons = [
+			{ action: 'cancel', label: translate( 'Cancel' ) },
+			{ action: 'submit', label: translate( 'Yes, Submit!' ), isPrimary: true },
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **101** times:
`translate( 'Submit' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by hewsut** [comment]
I have autocomplete list, I'm talking to the API backend via data layer, I have success/fail feedback on that call, I have AYS popover in place.

Remaining todo:
o Wire in a Redux attribute that we can check on success -- to fully adjust the UI to reflect and send user to gaining registrar.
o Fix dialog max width to something less sprawling. (50%? 60%?)

**Comment by drw158** [comment]
I don't know the full context, so just take this with a grain of salt!

- The Transfer Confirmation modal looks really wide. Is there a version of the component that is smaller in width?
- After you hit Yes, there is a brief pause before the notice appears. You may consider disabling the Submit button during that time. Ppl on slow connections might have to wait longer during the pause.
- Instead of a card with a left border, I think it should be a regular notice. If you do that, then maybe you could get rid of the global notice, since it's a bit redundant.

regular notice:
screen shot 2018-01-30 at 1 02 58 pm

**Comment by hewsut** [comment]
> The Transfer Confirmation modal looks really wide. Is there a version of the component that is smaller in width?

Yeah, totes agree, I was thinking like 50% or 60% width maybe. There's nothing prepackaged AFAIK to specifically adjust the width. I think I need to explicitly style it with CSS. Which...  mumble mumble... I'll figure out. The component docs spell out how to override the default stylings. I just don't wanna. ;)

> You may consider disabling the Submit button during that time. Ppl on slow connections might have to wait longer during the pause.

Good call.  

> Instead of a card with a left border, I think it should be a regular notice. If you do that, then maybe you could get rid of the global notice, since it's a bit redundant.

"card with a left border" = "Success! Click here to go to the new registrar" 
"global notice" = "IPS Tag save successful."

Correct?

File: client/state/selectors/get-gaining-registrar.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Returns gaining registrar data for a given domain, if we've successfully
+ * completed the transfer from our side.
+ *
+ * @param  {Object}  state       Global state tree
+ * @param  {string}  domain      Domain
+ * @return {Object}              Registrar object/record
+ */
+export default function getGainingRegistrar( state, domain ) {
+	return get( state, 'domains.transfer["' + domain + '"].selectedRegistrar', {} );
```
Comment (ramonjd) [inline client/state/selectors/get-gaining-registrar.js]: Feel free to ignore this, but if you get comma coma, you can always use template literals here:
```
get( state, `domains.transfer['${ domain }'].selectedRegistrar`, {} );
```

File: client/state/selectors/is-transfer-done.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Return a boolean value indicating whether a transfer for a given domain
+ * is confirmed completed on our side.
+ *
+ * @param  {Object} state  Global state tree
+ * @param  {string}  domain      Domain
+ * @return {Boolean} If the transfer is done
+ */
+export default function isTransferDone( state, domain ) {
+	return get( state, 'domains.transfer["' + domain + '"].completeOnOurSide', false );
```
Comment (ramonjd) [inline client/state/selectors/is-transfer-done.js]: As above

File: client/state/data-layer/wpcom/domains/transfer/index.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		)
+	);
+};
+
+export const handleIpsSaveSuccess = ( { dispatch }, action ) => {
+	const { domain, selectedRegistrar } = action;
+
+	dispatch( updateDomainTransfer( domain, selectedRegistrar ) );
+	dispatch(
+		successNotice( translate( 'IPS tag save succesful.' ), {
```
Comment (ramonjd) [inline client/state/data-layer/wpcom/domains/transfer/index.js]: successful

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+						{ args: { selectedDomainName }, components: { strong: <strong /> } }
+					) }
+				</p>
+
+				<SearchCard
+					disableAutocorrect
+					onSearch={ this.handleSearchInputChange }
+					onBlur={ this.hideSuggestions }
+					onKeyDown={ this.handleKeyDown }
+					placeholder="Start typing an IPS tag..."
```
Comment (ramonjd) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Should this be translated?

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+					value={ this.state.ipsTagInput }
+				/>
+				<Suggestions
+					ref={ this.setSuggestionsRef }
+					query={ this.state.currentQuery }
+					suggestions={ this.getSuggestions() }
+					suggest={ this.handleSuggestionClick }
+				/>
+				<FormButton onClick={ this.popOverDialog } >
+					{ this.props.translate( 'Submit' ) }
```
Comment (ramonjd) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: You're already deconstructing `translate` so we can dispense with `this.props`

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+		} );
+	}
+
+	onCloseDialog = ( action ) => {
+		if ( 'submit' === action ) {
+			this.handleSubmit();
+		}
+
+		this.setState( { showDialog: false } );
+	}
```
Comment (ramonjd) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Another 'if you want to' stylistic modification:

```
onCloseDialog = ( action ) => ( 'submit' === action ) ? this.handleSubmit() : this.setState( { showDialog: false } );
```

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+	state = {
+		currentQuery: '',
+		ipsTagInput: '',
+		ipsTagList: [],
+		selectedRegistrar: {},
+		showDialog: false,
+	};
+
+	componentWillMount() {
+		request.get( SelectIpsTag.ipsTagListUrl ).then( ( response ) => {
```
Comment (ramonjd) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: I think `componentDidMount` is preferred to load data from a remote endpoint. 

Pretty sure `componentWillMount` is called on server side rendering as well.

https://reactjs.org/docs/react-component.html#componentdidmount

Comment (hewsut) [reply client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Nice. TIL, thanks!

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx (hunk 9 above)
Comment (hewsut) [reply client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Woop, yep, artifact from old copy pasta.

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx (hunk 8 above)
Comment (hewsut) [reply client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Erm, yas.

File: client/state/selectors/get-gaining-registrar.js (hunk 5 above)
Comment (deBhal) [reply client/state/selectors/get-gaining-registrar.js]: As a 3rd alternative, get accepts an array - `get( state, [ 'domains', 'transfer', domain, 'selectedRegistrar' ], {} )`.

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+	}
+
+	renderGoToGainingRegistrar() {
+		const { translate, gainingRegistrar: { registrarUrl } } = this.props;
+
+		return (
+			<Card
+				className="transfer-out__select-ips-tag"
+				href={ registrarUrl }
+				target="_blank"
```
Comment (deBhal) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: `rel="noopener noreferrer"` - https://dev.to/ben/the-targetblank-vulnerability-by-example

**Comment by deBhal** [comment]
The suggestion UI looks great!

> Figure out what to do after the user sets the tag but before they have completed at gaining registrar. (Probably remove sub on our side.)

It'd be great to have the url available in case the user gets confused (which seems pretty likely).

It looks like we're storing the gaining registrar in the state, so the user can come back to this page to get the link, as long as it's in the same browser, is that right?

I see we're recording it in the logs too, so a happiness engineer can dig it up too. (The whois on `hew-privacy-test02.uk` still says tucows, so I think that's no help).

I was wondering about adding it to domains/manage, but then I changed my mind - that's quite a lot of effort for a problem that _might_ happen to someone some-day. I made a pretty picture of the idea though:
julesauspremiumtest_ _wordpress_com

**Comment by drw158** [comment]
@hewsut yep!

**Comment by hewsut** [comment]
@drw158 I'm happy to switch the global notice to a regular notice.

...on a scale of "I hate it" to "burn it with fire", how much do you hate the card?

Is it the odd color highlight? Or the strange actionable placement relative to Calypso norms? 

If the color, I'd be happy to quash the highlight. That's NBD. The latter...  bear with me:

- I loosely grok that we have actionable patterns established [usually w/notice as you suggest?].
- That makes sense--usually we're doing something on a page and we want to call something else out and leave it to the reader on whether to follow through or just stay.
- Here, because of the way the transfer works, IMHO we WANT to "take over" their experience to _strongly_ encourage them to follow through (off-site) right now.

Or something else?

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+			{ action: 'submit', label: translate( 'Yes, Submit!' ), isPrimary: true },
+		];
+
+		return (
+			<Dialog
+				baseClassName="transfer__ips_tag_set_dialog"
+				buttons={ buttons }
+				isVisible={ this.state.showDialog }
+				onClose={ this.onCloseDialog }>
+				<h1>{ translate( 'Transfer Confirmation' ) }</h1>
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer confirmation required' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+			</Dialog>
+		);
+	}
+
+	render() {
+		const { translate, saveStatus } = this.props;
+
+		return (
+			<div>
+				<SectionHeader label={ translate( 'Transfer Domain To Another Registrar' ) } />
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer to another registrar' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+			{ action: 'submit', label: translate( 'Yes, Submit!' ), isPrimary: true },
+		];
+
+		return (
+			<Dialog
+				baseClassName="transfer__ips_tag_set_dialog"
+				buttons={ buttons }
+				isVisible={ showDialog }
+				onClose={ this.onCloseDialog }>
+				<h1>{ translate( 'Transfer Confirmation' ) }</h1>
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Transfer confirmation required' )` *ES Score: 16*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/uk-transfer-out)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+			</div>
+		);
+	}
+
+	renderDialog() {
+		const { selectedDomainName, translate } = this.props;
+		const { selectedRegistrar, showDialog } = this.state;
+		const buttons = [
+			{ action: 'cancel', label: translate( 'Cancel' ) },
+			{ action: 'submit', label: translate( 'Yes, Submit!' ), isPrimary: true },
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-out/select-ips-tag.jsx]: Hi! I've found a possible matching string that has already been translated **101** times:
`translate( 'Submit' )` *ES Score: 6*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by hewsut** [comment]
@drw158 ignore me, I get it now. I thought regular notices only popped at top too but I see how they can be used. I've swapped a regular notice into place (see new video at top).

**Comment by ramonjd** [comment]
Tested according instructions provided in PR in FF and Chrome. Flow proceeds as expected 👍 

Only some minor UX aberrations in mobile, which do not affect functionality. Worthy of soft launch.

screen shot 2018-02-02 at 3 13 42 pm
screen shot 2018-02-02 at 3 20 56 pm

**Pull request merged**